[CEED] libCeed integration [yohann/okina-gpu-libceed]

YohannDudouit opened the pull request:
This branch proposes a way to integrate libCEED in the current master branch.

_Technical meeting remarks:_
- [x] guard CEED parts in `bilininteg.cpp`
- [x] Capitalise functions
- [x] Replace `Device::IsEnabled()` with `Device::Allows(CUDA)`
- [x] `AddMult` currently does `Mult`, exploring solutions
- [x] Remove dead code in `libceed/mass.cpp`
- [x] `MFEM_SOURCE_DIR` does not necessarily exists, use `MFEM_INSTALL_DIR` in that case
- [x] `CMake`

| PR | Author | Editor | Reviewers  | Assignment | Approval | Merge |
| --- | --- | --- | ---  | --- | --- | --- |
| https://github.com/mfem/mfem/pull/866 | @YohannDudouit | @tzanio | @v-dobrev + @camierjs | 11/23/19 | 12/23/19 | ⌛due 01/02/20 |

YohannDudouit commented:
@v-dobrev @camierjs @tzanio I believe this PR is ready for review.
I added the option `MFEM_USE_CEED`, and the device option are `ceed-cpu` and `ceed-gpu`.

tzanio commented on fem/libceed.hpp:
```diff
@@ -0,0 +1,64 @@
+typedef void* CeedData;
+#endif
+
+}
```
Add newlines

tzanio commented on fem/libceed.okl:
```diff
@@ -0,0 +1,245 @@
+// *****************************************************************************
```
Add MFEM header

YohannDudouit commented on general/ceed.hpp:
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
```
I created this file to mimic what was done for Occa and Cuda, it was convenient to have it, but I'm open to any proposition to remove it.

YohannDudouit commented on general/device.cpp:
```diff
@@ -28,20 +29,21 @@ CUstream *cuStream = NULL;
-   "occa-cpu", "raja-cpu", "cpu"
+   "occa-cuda", "raja-cuda", "ceed-cuda", "cuda", "occa-omp", "raja-omp", "omp",
+   "occa-cpu", "raja-cpu", "ceed-cpu", "cpu"
 };
```
The current priority is hardcoded in MFEM, there isn't any way to quickly change the priority for now. Did you have some priority requests?
Should I add a 4.1 section in the CHANGELOG?

v-dobrev replied:
I updated this -- libCEED now has the highest priority for both the CPU and the CUDA backends when configured/enabled.

YohannDudouit commented:
@v-dobrev @tzanio @camierjs The integration of libCEED has been successfully rebased, it can be reviewed.

YohannDudouit commented:
After some investigation, it seems that there is some unexpected transfers between host and device. I'll remove the WIP once I fixed that.

YohannDudouit commented:
The bug is in libCEED.

tzanio commented:
@YohannDudouit is this ready for review? Can you fix the Appveyor errors?

I'd like to merge this before CEED3AM.

YohannDudouit commented:
@tzanio The problem came from the metis server that was down, @v-dobrev resolved the problem.

YohannDudouit commented:
@tzanio I added the support for the `cuda-gen` backend in this PR since it has been merged in the master branch of libCEED.

YohannDudouit commented:
@tzanio Should I make the `cuda-gen` backend the default libCEED gpu backend for mfem?

tzanio commented:
> @tzanio Should I make the cuda-gen backend the default libCEED gpu backend for mfem?

Yes

camierjs commented:
There are specific `diffusion` and `mass` code in the `libceed.cpp` file: couldn't we separate the integrators to the `libceed` generic code?

v-dobrev commented:
The backend priorities in the doxygen documentation of `Device::Configure` need to be updated -- libceed is not listed there.

tzanio commented on general/device.hpp:
```diff
@@ -179,7 +189,13 @@ class Device
        * Multiple backends can be configured at the same time.
```
Update the backend priority list in the documentation of `Device:: Configure`?

v-dobrev replied:
Done.

tzanio commented on general/ceed.hpp:
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
```
I suggest we remove this file and maybe use `libceed.hpp` instead.

tzanio commented on fem/bilinearform_ext.cpp:
```diff
@@ -102,23 +102,33 @@ void PABilinearFormExtension::Mult(const Vector &x, Vector &y) const
       {
-         integrators[i]->AddMultPA(localX, localY);
+         integrators[i]->AddMultPA(x, y);
       }
```
Does this work in parallel? What about with AMR?

YohannDudouit replied:
I believe that this work in parallel, `local` here means that it is an E-vector, but I let libCEED do the scatter/gather, that's why I'm giving the L-vectors `x` and `y`.
AMR doesn't work since it is not yet supported by libCEED, I imagine that if all we need is a weight vector for AMR, then it would be straightforward to add AMR to libCEED.

v-dobrev replied:
I think this works fine even with AMR and in parallel since both `x` and `y` are `GridFunction`-size vectors and the transition to and from E-vector is still a boolean matrix even in the case of AMR. This means that the scatter and gather in libCEED are sufficient. The real parallel and/or AMR interpolation is in the prolongation and restriction operators applied on top of this `Mult` action. These prolongation/restriction operators are defined in the base class `BilinearFormExtension` by the methods `GetProlongation` and `GetRestriction` which just use the respective methods from the internally stored `BilinearForm` pointer `a`.

YohannDudouit replied:
I tried it, AMR works.

tzanio commented on fem/libceed_mass.cu:
```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
```
I suggest we reorganize these files as follows:
- make a subdirectory `fem/libceed`
- `libceed.*` -> `libceed/ceed.*`
- `libceed_mass.*` -> `libceed/mass.*`
- `libceed_diffusion.*` -> `libceed/diffusion.*`

YohannDudouit commented:
@v-dobrev Rebased with master and updated to the last libCEED version.
I had to define a bunch of `CEED` stuff to make it work when `MFEM_USE_CEED=NO`, alternatively we could create a `qfunction` subdirectory to avoid issues with libCEED qFunctions.

barker29 commented on fem/libceed/mass.h:
```diff
@@ -0,0 +1,106 @@
+   const CeedScalar coeff = bc->coeff;
+   const CeedScalar *J = in[0], *qw = in[1];
+   CeedScalar *rho = out[0];
+   switch (bc->dim + 10*bc->space_dim)
```
Ditto.

barker29 commented on fem/libceed/mass.h:
```diff
@@ -0,0 +1,106 @@
+   BuildContext *bc = (BuildContext *)ctx;
+   const CeedScalar *c = in[0], *J = in[1], *qw = in[2];
+   CeedScalar *rho = out[0];
+   switch (bc->dim + 10*bc->space_dim)
```
Ditto.

YohannDudouit commented on fem/libceed/diffusion.h:
```diff
@@ -0,0 +1,199 @@
+   const CeedScalar coeff = bc->coeff;
+   const CeedScalar *J = in[0], *qw = in[1];
+   CeedScalar *qd = out[0];
+   switch (bc->dim + 10 * bc->space_dim)
```
libCEED doesn't currently support an unified error system for CPU and GPU. This might get added later.

YohannDudouit replied:
To be more precise, there use to be an error message but it has been removed in the last version of libCEED for compatibility with GPUs.

barker29 replied:
There's nothing you can do? `std::abort` or `exit(1)` or `mfem_error()` or something? As it is this will just execute silently and the user will have no idea what is going on.

v-dobrev replied:
I think the issue is that this code can be turned into a kernel on the GPU and then calling CPU functions is not possible. One solution could be to perform the check outside of this kernel and just leave a comment here that the check is in the calling scope.

artv3 replied:
Another possible option is to use ``#if defined(__CUDA__ARCH__)`` to detect if the code will be offloaded to the GPU. If so, you can use  ``asm ("trap;")`` to halt the kernel. Or abort() on the cpu. 
There is a nice solution here: 
https://github.com/LLNL/RAJA/blob/develop/include/RAJA/config.hpp.in#L205

YohannDudouit replied:
You only know at runtime if the code is offloaded on GPU, I don't think that this compilation time solution would work for libCEED.

camierjs replied:
Couldn't we use the `ctx` to get this info back?

YohannDudouit replied:
This is using the `ctx`

YohannDudouit commented on fem/bilininteg.hpp:
```diff
@@ -1738,6 +1740,8 @@ class MassIntegrator: public BilinearFormIntegrator
    const GeometricFactors *geom;  ///< Not owned
    int dim, ne, nq, dofs1D, quad1D;
+   // CEED extension
+   void* ceedDataPtr;
```
This is a `void*` because there will be different `CeedData`, one for PA, one for MF.

YohannDudouit replied:
See #1101

barker29 replied:
So I've looked at #1101, and as far as I can tell this is always a `CeedData*`, for example in both `MassIntegrator` and `DiffusionIntegrator` methods `::AddMultPA` and `::AddMultMF`, it's always a `CeedData*`, so I'm not seeing the reason for `void *`.

YohannDudouit replied:
You're correct, I had planned to use different `CeedData` for `PA` or `MF`, but it was troublesome, I'll do this modification.

YohannDudouit commented on fem/bilininteg_diffusion.cpp:
```diff
@@ -1088,9 +1101,41 @@ static void PADiffusionApply(const int dim,
+         y_ptr = y.HostReadWrite();
+         mem = CEED_MEM_HOST;
+      }
+      CeedData& ceedData = *static_cast<CeedData*>(ceedDataPtr);
```
With MF Ceed operator I won't be able to avoid it.

YohannDudouit replied:
See #1101

YohannDudouit commented on fem/bilininteg_mass.cpp:
```diff
@@ -783,7 +796,39 @@ static void PAMassApply(const int dim,
+         y_ptr = y.HostReadWrite();
+         mem = CEED_MEM_HOST;
+      }
+      CeedData& ceedData = *static_cast<CeedData*>(ceedDataPtr);
```
See #1101

YohannDudouit commented on fem/libceed/ceed.hpp:
```diff
@@ -0,0 +1,81 @@
+
+// struct BuildContextConstCoeff { CeedInt dim, space_dim; CeedScalar coeff; };
+
+enum CeedCoeff { Const, Grid };
```
@v-dobrev Any opinion?

v-dobrev replied:
We already have `enum class` in a few places. I also think `enum class` is better here, otherwise `Const` and `Grid` are too generic.

YohannDudouit commented on fem/libceed/ceed.hpp:
```diff
@@ -0,0 +1,81 @@
+
+void initCeedCoeff(Coefficient* Q, CeedData* ptr);
+
+void FESpace2Ceed(const mfem::FiniteElementSpace &fes,
```
My bad, I changed it in #1100 and #1101 , I'll propagate the change to this PR too.

YohannDudouit replied:
Tell me if you find the new name self documented, or if you think I should still add a documentation.

barker29 replied:
The name is better.

jeremylt commented on general/device.cpp:
```diff
@@ -251,6 +288,28 @@ void Device::Setup(const int device)
+   {
+      if (!ceed_option)
+      {
+         CeedDeviceSetup("/cpu/self/opt/blocked");
```
I just thought of this - if this is set to `/cpu/self/`, then libCEED will pick the best available CPU backend (so LIBXSMM > AVX > Opt > Ref)

YohannDudouit replied:
Thank's that's a really good point!

camierjs commented on fem/libceed_diffusion.qf:
```diff
@@ -0,0 +1,174 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
```
Could we move these `qf` files to the `ceed` subdir?

YohannDudouit replied:
That's a mistake, I should have removed them.

camierjs commented on fem/libceed_mass.qf:
```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
```
Same thing, should be in the `ceed` subdir.

YohannDudouit replied:
I'll remove this file too.

camierjs commented:
The `nvvp` traces for multiple integrators are really great!
What are these `SetValueK`?
![ceed](https://user-images.githubusercontent.com/20000517/67589069-75d3e500-f70c-11e9-873c-4370a362fe4d.png)

YohannDudouit commented:
`SetValueK` is a Kernel that sets a value in a `CeedVector`, it is mostly used to set the vectors to zero.

camierjs commented:
No ways to pass the info without a kernel launch when it's just a set-to-zero?

vladotomov commented on fem/libceed/diffusion.cpp:
```diff
@@ -0,0 +1,142 @@
+{
+
+#ifdef MFEM_USE_CEED
+
```
Why are there separate codes (in diffusion.cpp and mass.cpp) enclosing the kernels for mass and diffusion? I would expect computing some values at quad points to be conceptually the same, up to the kernels. The two files mass.cpp and diffusion.cpp two are nearly identical - can't they be combined?

If not, do you expect app developers to write a function like this for every custom integrator, only to compute the quad data? And this doesn't include the actual kernels.

YohannDudouit replied:
It's true that these codes are largely similar, I was planning on factoring them in the future, when I would write a PR for vectorial mass and diffusion.

vladotomov commented on fem/libceed/diffusion.cpp:
```diff
@@ -0,0 +1,142 @@
+                        ceedData.basis, CEED_VECTOR_ACTIVE);
+
+   CeedVectorCreate(ceed, fes.GetNDofs(), &ceedData.u);
+   CeedVectorCreate(ceed, fes.GetNDofs(), &ceedData.v);
```
Sorry to be so much behind, but why are 32 Ceed* function calls needed to compute 1D array of values? Does this have to be in MFEM and who will do it for the application-specific integrators?

YohannDudouit replied:
libCEED is a bit verbose, however I'm not following the comment about 1D array of values, this function creates two operators, one to setup and one to apply the diffusion integrator, it also applies the setup operator.
As long as we don't have some sort of UFL, we can't generate this kind of code automatically, thus each integrator as to be written by hand by someone.

vladotomov commented on fem/libceed/diffusion.h:
```diff
@@ -0,0 +1,199 @@
+/// A structure used to pass additional data to f_build_diff and f_apply_diff
+struct BuildContext { CeedInt dim, space_dim; CeedScalar coeff; };
+
+/// libCEED Q-function for building quadrature data for a diffusion operator
```
Can't these Q-functions reuse code somehow? The two functions in this file are nearly identical.

YohannDudouit replied:
I don't see how, but if you have a proposition I would be interested to know.

YohannDudouit replied:
I think using the `CEED_QFUNCTION` macro would work to create functions, however I don't really see how to factorize code.

v-dobrev commented:
Running the following example:
```console
./ex6 -pa -d ceed-cuda -no-vis
```
takes a lot longer than
```console
./ex6 -pa -d cuda -no-vis
```
It will be good to figure out what the reason for that is.

v-dobrev commented on fem/bilininteg.hpp:
```diff
@@ -1679,6 +1683,11 @@ class DiffusionIntegrator: public BilinearFormIntegrator
 
+#ifdef MFEM_USE_CEED
+   // CEED extension
+   CeedData* ceedDataPtr;
```
It seems like this struct is never destroyed. Am I missing something?

v-dobrev replied:
It seem like a good idea to add a constructor (initializing an empty object) and a destructor for `CeedData`.

YohannDudouit commented:
@v-dobrev and @camierjs I think I have addressed all your concerns for this PR, the `CeedData` doesn't leak memory any more and I replaced `CeedOperatorApply` with `CeedOperatorApplyAdd`.

camierjs commented:
With a fresh `libCEED`/`master` installation, I'm hitting the following issue:
`bilininteg.hpp(1685): error: identifier "CeedData" is undefined`
Have you seen this before?

YohannDudouit commented:
@camierjs Sorry for this issue with the circular dependency, it compiled previously, but I must have forgotten to make a `make clean` before, this should work now.

tzanio commented:
Reminder: the reviews for this PR are due this week

cc: @v-dobrev, @camierjs

camierjs commented:
@YohannDudouit, I still have issues with `make distclean && make serial -j MFEM_USE_CEED=YES && make test -j`: tests `unit_tests`, `ex8`, `ex16`, `ex18`, `ex19`, `NURBS ex1` are failing, and more with the parallel build.

YohannDudouit commented:
@camierjs All tests now passes with `make distclean && make serial -j MFEM_USE_CEED=YES && make test -j`

YohannDudouit commented:
@camierjs All tests pass in parallel too.

YohannDudouit commented:
@camierjs The sample runs pass for ceed-cpu and ceed-cuda (despite the `AtomicAdd`):
```
Summary:
--------
Examples:
[  OK  ] ex1 -pa -d cuda -no-vis
[FAILED] ex1 -pa -d raja-cuda -no-vis
[FAILED] ex1 -pa -d occa-cuda -no-vis
[FAILED] ex1 -pa -d raja-omp -no-vis
[FAILED] ex1 -pa -d occa-omp -no-vis
[  OK  ] ex1 -pa -d ceed-cpu -no-vis
[  OK  ] ex1 -pa -d ceed-cuda -no-vis
[  OK  ] ex1 -m ../data/beam-hex.mesh -pa -d cuda -no-vis
[  OK  ] ex6 -pa -d cuda -no-vis
[FAILED] ex6 -pa -d occa-cuda -no-vis
[FAILED] ex6 -pa -d raja-omp -no-vis
[  OK  ] ex6 -pa -d ceed-cpu -no-vis
[  OK  ] ex6 -pa -d ceed-cuda -no-vis
```
PS: occa and raja fail because they are not installed.

camierjs commented:
`valgrind ./ex1 -pa -d ceed-cpu:/cpu/self/ref/serial` 

hits a 
```vex amd64->IR: unhandled instruction bytes: 0x62 0xA1 0x65 0x40 0xEF 0xDB 0x62 0xE1 0xFF 0x8
vex amd64->IR:   REX=0 REX.W=0 REX.R=0 REX.X=0 REX.B=0
vex amd64->IR:   VEX=0 VEX.L=0 VEX.nVVVV=0x0 ESC=NONE
vex amd64->IR:   PFX.66=0 PFX.F2=0 PFX.F3=0
==202143== valgrind: Unrecognised instruction at address 0x4e42920.
```
We might want to forward this to `libCEED`, as it has the same issue.

v-dobrev commented:
@YohannDudouit, I updated our LLNL testing and the script `config/sample-runs.sh` to support libCEED -- now it should be fine to add sample runs using `ceed-*` devices (from your previous comments I see that you already did that locally on your machine).

jeremylt commented:
@camierjs, can you provide the details on the Valgrind issue? Our pure libCEED examples are Valgrind clean.

tzanio commented:
There is a merge conflict now because of #1107. Can you address it?

camierjs commented:
On Lassen (`gcc/7.3.1` `cuda/10.1.243`) with a `libCEED` build w/o `occa`, in `ex1` with the `/cpu/self/ref/serial` backend, we still get a lot of `Conditional jump or move depends on uninitialised value(s)`, `Invalid read` and one `block definitely lost`:  [ex1.txt](https://github.com/mfem/mfem/files/3974719/ex1.txt)

jeremylt commented:
It looks like some sort of a problem with the strings being passed to libCEED, as far as I can tell.

tzanio commented:
Pushed target approval date by one week to 12/21/19

v-dobrev commented:
I still see memory leaks with valgrind but they seem to be outside of MFEM.

For example, one leak in libCEED is here: `ceed-cuda-gen-qfunction.c:82` and another is here: `ceed-cuda-gen-qfunction.c:70`. Valgrind also shows invalid reads at `ceed-cuda.c:58` and `ceed-cuda-gen-operator-build.cpp:835` -- both access data outside of an allocated block of size 0.

tzanio commented:
The regression testing on tux429, see [mfem/autotes/tux429/2019-12-26-github-next/sample-runs-devices.diff](https://mybitbucket.llnl.gov/projects/MFEM/repos/autotest/commits/bb8d10e0b07e453902557917a20b4f15e4d4e698#tux429/2019-12-26-github-next/sample-runs-devices-tux429.diff), found the following error which seems to be a conflict between this PR and #982.

Can you please review and fix?

```
<examples> ex1 -pa -d ceed-cpu -no-vis
----------------------------------------------------------------
Options used:
   --mesh ../data/star.mesh
   --order 1
   --no-static-condensation
   --partial-assembly
   --device ceed-cpu
   --no-visualization
Device configuration: ceed-cpu,cpu
libCEED backend: /cpu/self/avx/blocked
Number of finite element unknowns: 20801
Size of linear system: 20801


MFEM abort: Unknown kernel.
 ... in function: PADiffusionAssembleDiagonal
 ... in file: fem/bilininteg_diffusion.cpp:671
``` 

(There are similar errors for `ex1 -pa -d ceed-cuda -no-vis`, `mpirun -np 4 ex1p -pa -d ceed-cpu -no-vis`, and `mpirun -np 4 ex1p -pa -d ceed-cuda -no-vis`.)

tzanio commented:
With the merge of #982, the above is an error. Can you please address?

YohannDudouit commented:
@tzanio I believe I resolved the issue.

tzanio commented:
Re-merged in `next` for testing ...

tzanio commented:
@camierjs and @v-dobrev -- if I don't hear from you, I'll assume you still approve

v-dobrev commented:
Looks good to me. I did not test the latest commit -- we'll see test results from `next` ...

camierjs commented:
Rerun some tests successfully.

tzanio commented:
This is a major milestone for MFEM and libCEED that will be of great benefit to both projects and their users.

@YohannDudouit, thank you for your hard work and perseverance on this PR!

@camierjs, @v-dobrev and @jeremylt, thank you for your help reviewing and improving the code!

Congratulations 🏆

tzanio commented:
The regression tests on tux429 show [non-deterministic behaviour in Example6p](https://mybitbucket.llnl.gov/projects/MFEM/repos/autotest/commits/c6aec8bc5bac20662f3475508ddadfda84184aef#tux429/2020-01-12-github-master/sample-runs-devices-tux429.diff): see the diff below between two consecutive runs.

@YohannDudouit and @camierjs, can you please investigate and fix?

```diff
--- sample-runs-devices-tux429.saved.filt	2020-01-12 01:46:36.178279017 -0800
+++ sample-runs-devices-tux429.out.filt	2020-01-12 01:46:36.150278875 -0800
@@ -2003,56 +2003,56 @@ <examples> ex6 -pa -d ceed-cuda -no-vis
 AMR iteration 11
 Number of unknowns: 3331
    Iteration :   0  (B r, r) = 0.000849051 ...
-   Iteration : 110  (B r, r) = 7.65147e-16
-Average reduction factor = 0.881554
+   Iteration : 110  (B r, r) = 7.17117e-16
+Average reduction factor = 0.881295
 
 AMR iteration 12
 Number of unknowns: 4176
    Iteration :   0  (B r, r) = 0.000323304 ...
-   Iteration : 123  (B r, r) = 2.60116e-16
-Average reduction factor = 0.892967
+   Iteration : 123  (B r, r) = 2.6025e-16
+Average reduction factor = 0.892969
 
 AMR iteration 13
 Number of unknowns: 5816
    Iteration :   0  (B r, r) = 0.000365596 ...
-   Iteration : 143  (B r, r) = 3.04415e-16
-Average reduction factor = 0.907327
+   Iteration : 143  (B r, r) = 3.19342e-16
+Average reduction factor = 0.907479
 
 AMR iteration 14
 Number of unknowns: 10521
    Iteration :   0  (B r, r) = 0.000310254 ...
-   Iteration : 215  (B r, r) = 2.63238e-16
+   Iteration : 215  (B r, r) = 2.6331e-16
 Average reduction factor = 0.937405
 
 AMR iteration 15
 Number of unknowns: 13956
    Iteration :   0  (B r, r) = 0.00013188 ...
-   Iteration : 244  (B r, r) = 1.12195e-16
-Average reduction factor = 0.944639
+   Iteration : 243  (B r, r) = 1.26393e-16
+Average reduction factor = 0.944649
 
 AMR iteration 16
 Number of unknowns: 17956
    Iteration :   0  (B r, r) = 9.17753e-05 ...
-   Iteration : 270  (B r, r) = 6.93591e-17
-Average reduction factor = 0.949626
+   Iteration : 270  (B r, r) = 6.93387e-17
+Average reduction factor = 0.949625
 
 AMR iteration 17
 Number of unknowns: 24256
    Iteration :   0  (B r, r) = 8.7984e-05 ...
-   Iteration : 336  (B r, r) = 6.52734e-17
-Average reduction factor = 0.95929
+   Iteration : 334  (B r, r) = 8.38041e-17
+Average reduction factor = 0.95941
 
 AMR iteration 18
 Number of unknowns: 43969
    Iteration :   0  (B r, r) = 6.7787e-05 ...
-   Iteration : 501  (B r, r) = 5.66363e-17
-Average reduction factor = 0.972626
+   Iteration : 501  (B r, r) = 5.93267e-17
+Average reduction factor = 0.972671
 
 AMR iteration 19
 Number of unknowns: 56668
    Iteration :   0  (B r, r) = 3.11646e-05 ...
-   Iteration : 609  (B r, r) = 2.94222e-17
-Average reduction factor = 0.977524
+   Iteration : 609  (B r, r) = 2.97023e-17
+Average reduction factor = 0.977531
 Reached the maximum number of dofs. Stop.
 [  OK  ] <examples> ex6 -pa -d ceed-cuda -no-vis
 
@@ -10645,7 +10645,7 @@ <examples> mpirun -np 4 ex6p -pa -d ceed
 AMR iteration 22
 Number of unknowns: 83501
    Iteration :   0  (B r, r) = 7.61381e-05 ...
-   Iteration : 264  (B r, r) = 7.15021e-17
+   Iteration : 264  (B r, r) = 7.1483e-17
 Average reduction factor = 0.948901
 
 AMR iteration 23
@@ -10657,8 +10657,8 @@ <examples> mpirun -np 4 ex6p -pa -d ceed
 AMR iteration 24
 Number of unknowns: 86266
    Iteration :   0  (B r, r) = 4.4325e-06 ...
-   Iteration : 277  (B r, r) = 3.85466e-18
-Average reduction factor = 0.951108
+   Iteration : 277  (B r, r) = 3.85804e-18
+Average reduction factor = 0.95111
 
 AMR iteration 25
 Number of unknowns: 91806
@@ -10669,8 +10669,8 @@ <examples> mpirun -np 4 ex6p -pa -d ceed
 AMR iteration 26
 Number of unknowns: 123036
    Iteration :   0  (B r, r) = 7.30948e-06 ...
-   Iteration : 314  (B r, r) = 6.80396e-18
-Average reduction factor = 0.956846
+   Iteration : 314  (B r, r) = 7.26175e-18
+Average reduction factor = 0.956945
 Reached the maximum number of dofs. Stop.
 [  OK  ] <examples> mpirun -np 4 ex6p -pa -d ceed-cuda -no-vis
```
